endpoint/OBDComp/monitor/monitor.c

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "interfaces.h"
+#include "legato.h"
+
+#include "brotli/decode.h"
+#include "brotli/encode.h"
+#include "build/obd_generated.h"
+#include "cipher.h"
+#include "endpoint/OBDComp/can-bus/can-utils.h"
+#include "endpoint/OBDComp/obd_pid.h"
+#include "hal/device.h"
+
+#include "legato.h"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int string_compress(uint8_t *origin, size_t len, uint8_t **out_buffer, size_t *out_size) {
+  int lgwin = BROTLI_DEFAULT_WINDOW;
+  int bufferSize = len;
+  uint8_t *buffer = malloc(bufferSize);
+  uint8_t *inBuffer = malloc(bufferSize);
+
+  memcpy(inBuffer, origin, len);
+
+  int quality = 11;
+  size_t encodedSize = bufferSize;
+
+  bool brotliStatus =
+      BrotliEncoderCompress(quality, lgwin, BROTLI_MODE_GENERIC, len, (const uint8_t *)inBuffer, &encodedSize, buffer);
+
+  *out_buffer = buffer;
+  *out_size = encodedSize;
+
+  free(inBuffer);
+  return brotliStatus;
+}
+
+#undef ns
+#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(OBD, x)  // Specified in the schema.
+
+static uint32_t hex_to_uint32(uint8_t *arr, int offset, size_t len) {
```

Comment:
Why didn't you use `strtol` instead?

Reply:
1. arr is an uint8_t `array`, not string.
2. We only need to send the last 4 bytes of data. The name of the function may be misleading, I will change it.

Reply:
An `uint8_t` array can be converted to C-style string once you set the terminator.